litex/soc/software/bios/cmds/cmd_bios.c

```diff
@@ -8,10 +8,61 @@
 #include <sim_debug.h>
 
 #include <generated/csr.h>
-
+#include <generated/mem.h>
 #include "../command.h"
 #include "../helpers.h"
 
+
+#define DOT_VAL	0x80
+
+static unsigned char hex_pos[8] = {
+	0x01,
+	0x02,
+	0x04,
+	0x08,
+	0x10,
+	0x20,
+	0x40,
+	0x80
+};
+
+static unsigned char hex_val[16]  = {
+	(0x01 | 0x02 | 0x04 | 0x08 | 0x10 | 0x20),        // 0
+	(0x02 | 0x04),                                    // 1
+	(0x01 | 0x02 | 0x40 | 0x10 | 0x08),               // 2
+	(0x01 | 0x02 | 0x04 | 0x08 | 0x40),               // 3
+	(0x02 | 0x04 | 0x20 | 0x40),                      // 4
+	(0x01 | 0x04 | 0x20 | 0x40 | 0x08),               // 5
+	(0x01 | 0x04 | 0x08 | 0x10 | 0x20 | 0x40),        // 6
+	(0x20 | 0x01 | 0x02 | 0x04),                      // 7
+	(0x01 | 0x02 | 0x04 | 0x08 | 0x10 | 0x20 | 0x40), // 8
+	(0x01 | 0x02 | 0x04 | 0x08 | 0x20 | 0x40),        // 9
+	(0x01 | 0x02 | 0x04 | 0x10 | 0x20 | 0x40),        // A
+	(0x04 | 0x08 | 0x10 | 0x20 | 0x40),               // B
+	(0x01 | 0x08 | 0x10 | 0x20),                      // C
+	(0x02 | 0x04 | 0x08 | 0x10 | 0x40),               // D
+	(0x01 | 0x08 | 0x10 | 0x20 | 0x40),               // E
+	(0x01 | 0x10 | 0x20 | 0x40),                      // F
+};
+
+volatile uint32_t ticks_per_delay = 8;
+volatile uint32_t ticks_per_us = 100;
+
+static void delay(unsigned int counts) {	// ticks based
+	unsigned int i = counts;
+	while(i--) {__asm__ volatile("nop");}
+}
+
+static void delay_ms(unsigned int counts) {	// mini second based
+	unsigned int i = (1000 * counts * ticks_per_delay) / ticks_per_us;
+	while(i--) {__asm__ volatile("nop");}
+}
+
+static void delay_us(unsigned int counts) {	// micro second based
+	unsigned int i = (counts * ticks_per_delay) / ticks_per_us;
+	while(i--) {__asm__ volatile("nop");}
+}
+
 /**
```

Comment:
We already have `busy_wait` and `cdelay` functions in the BIOS, we should try to avoid introducing others unless if really required.